mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -471,6 +471,202 @@ void CastOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.add<ChainedTensorCast, TensorCastExtractSlice>(context);
 }
 
+//===----------------------------------------------------------------------===//
+// ConcatOp
+//===----------------------------------------------------------------------===//
+
+FailureOr<RankedTensorType> ConcatOp::inferResultType(int64_t dim,
+                                                      TypeRange inputTypes) {
+  if (dim < 0)
+    return failure();
```

Comment:
It seems un some of these cases, you're trying a bit too hard to be nice and end up duplicating the verifier.
Maybe we can turn some of these failures into assertions (but being very careful that they cannot be hit from verifier because that is another can of worms..)

Reply:
+1. I think we can move the checks to verifier and make this method return `RankedTensorType` type. Like other static methods, user should be careful when they use the util. It is okay to not return a failure in this static method. Adding assertions is good enough to me.